src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -216,6 +365,208 @@ DropOrphanedShards(bool waitForLocks)
+								 currentOperationRecordList))));
+
+		if (failedShardCountOnComplete > 0)
+		{
+			ereport(WARNING, (errmsg("Failed to drop %d cleanup shards out of %d",
+									 failedShardCountOnComplete, list_length(
+										 currentOperationRecordList))));
+		}
+	}
+}
+
+
+/*
+ * InsertCleanupRecordInCurrentTransaction inserts a new pg_dist_cleanup_record entry
+ * as part of the current transaction. This is primarily useful for deferred drop scenarios,
+ * since these records would roll back in case of operation failure.
+ */
+void
+InsertCleanupRecordInCurrentTransaction(CleanupObject objectType,
+										char *objectName,
+										int nodeGroupId,
+										CleanupPolicy policy)
+{
+	/* We must have a valid OperationId. Any operation requring cleanup
+	 * will call StartNewOperationNeedingCleanup.
+	 */
+	Assert(CurrentOperationId != INVALID_OPERATION_ID);
+
+	Datum values[Natts_pg_dist_cleanup];
+	bool isNulls[Natts_pg_dist_cleanup];
+
+	/* form new shard tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	uint64 recordId = GetNextCleanupRecordId();
+	OperationId operationId = CurrentOperationId;
+
+	values[Anum_pg_dist_cleanup_record_id - 1] = UInt64GetDatum(recordId);
+	values[Anum_pg_dist_cleanup_operation_id - 1] = UInt64GetDatum(operationId);
+	values[Anum_pg_dist_cleanup_object_type - 1] = Int32GetDatum(objectType);
+	values[Anum_pg_dist_cleanup_object_name - 1] = CStringGetTextDatum(objectName);
+	values[Anum_pg_dist_cleanup_node_group_id - 1] = Int32GetDatum(nodeGroupId);
+	values[Anum_pg_dist_cleanup_policy_type - 1] = Int32GetDatum(policy);
+
+	/* open cleanup relation and insert new tuple */
+	Oid relationId = DistCleanupRelationId();
+	Relation pgDistCleanup = table_open(relationId, RowExclusiveLock);
+
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCleanup);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCleanup, heapTuple);
+
+	CommandCounterIncrement();
+	table_close(pgDistCleanup, NoLock);
+}
+
+
+/*
+ * InsertCleanupRecordInSeparateTransaction inserts a new pg_dist_cleanup_record entry
+ * in a separate transaction to ensure the record persists after rollback. We should
+ * delete these records if the operation completes successfully.
+ *
+ * For failure scenarios, use a subtransaction (direct insert via localhost).
+ */
+void
+InsertCleanupRecordInSubtransaction(CleanupObject objectType,
+									char *objectName,
+									int nodeGroupId,
+									CleanupPolicy policy)
+{
+	/* We must have a valid OperationId. Any operation requring cleanup
+	 * will call StartNewOperationNeedingCleanup.
+	 */
+	Assert(CurrentOperationId != INVALID_OPERATION_ID);
+
+	StringInfo command = makeStringInfo();
+	appendStringInfo(command,
+					 "INSERT INTO %s.%s "
+					 " (record_id, operation_id, object_type, object_name, node_group_id, policy_type) "
+					 " VALUES ( nextval('%s'), " UINT64_FORMAT ", %d, %s, %d, %d)",
```

Comment:
minor: nextval -> pg_catalog.nextval, maybe also use the fully-qualified sequence name

Reply:
Fixed.